File: Python/matrix/parallel_matrix_multiplication.py
import numpy as np
from concurrent.futures import ThreadPoolExecutor

class ParallelMatrixMultiplication:
    def __init__(self, num_threads=4):
        self.num_threads = num_threads

    def multiply(self, matrix_a, matrix_b):
        rows = len(matrix_a)
        cols = len(matrix_b[0])
        common = len(matrix_b)
        result = np.zeros((rows, cols))
        transposed_b = self.transpose(matrix_b)

        with ThreadPoolExecutor(max_workers=self.num_threads) as executor:
            tasks = []
            self._create_tasks(matrix_a, transposed_b, result, 0, rows, 0, cols, common, executor, tasks)
            for task in tasks:
                task.result()

        return result

    @staticmethod
    def transpose(matrix):
        return np.array(matrix).T

    def _create_tasks(self, matrix_a, matrix_b, result, start_row, end_row, start_col, end_col, common, executor, tasks, threshold=256):
        if (end_row - start_row) * (end_col - start_col) <= threshold:
            tasks.append(executor.submit(self._compute, matrix_a, matrix_b, result, start_row, end_row, start_col, end_col, common))
        else:
            mid_row = (start_row + end_row) // 2
            mid_col = (start_col + end_col) // 2

            self._create_tasks(matrix_a, matrix_b, result, start_row, mid_row, start_col, mid_col, common, executor, tasks)
            self._create_tasks(matrix_a, matrix_b, result, start_row, mid_row, mid_col, end_col, common, executor, tasks)
            self._create_tasks(matrix_a, matrix_b, result, mid_row, end_row, start_col, mid_col, common, executor, tasks)
            self._create_tasks(matrix_a, matrix_b, result, mid_row, end_row, mid_col, end_col, common, executor, tasks)

    @staticmethod
    def _compute(matrix_a, matrix_b, result, start_row, end_row, start_col, end_col, common):
        for i in range(start_row, end_row):
            for j in range(start_col, end_col):
                result[i][j] = np.dot(matrix_a[i, :common], matrix_b[j, :common])

File: Java/src/main/java/org/ulpgc/matrix/parallel/ParallelMatrixMultiplication.java
package org.ulpgc.matrix.parallel;

import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.RecursiveAction;

public class ParallelMatrixMultiplication {

    private final ForkJoinPool pool;

    public ParallelMatrixMultiplication(int numThreads) {
        this.pool = new ForkJoinPool(numThreads);
    }

    public double[][] multiply(double[][] matrixA, double[][] matrixB) {
        int rows = matrixA.length;
        int cols = matrixB[0].length;
        int common = matrixB.length;
        double[][] result = new double[rows][cols];
        double[][] transposedB = transpose(matrixB);

        pool.invoke(new MultiplyTask(matrixA, transposedB, result, 0, rows, 0, cols, common));
        return result;
    }

    private static double[][] transpose(double[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        double[][] transposed = new double[cols][rows];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                transposed[j][i] = matrix[i][j];
            }
        }
        return transposed;
    }

    private static class MultiplyTask extends RecursiveAction {
        private final double[][] matrixA;
        private final double[][] matrixB;
        private final double[][] result;
        private final int startRow, endRow, startCol, endCol, common;

        private static final int THRESHOLD = 64;

        MultiplyTask(double[][] matrixA, double[][] matrixB, double[][] result,
                     int startRow, int endRow, int startCol, int endCol, int common) {
            this.matrixA = matrixA;
            this.matrixB = matrixB;
            this.result = result;
            this.startRow = startRow;
            this.endRow = endRow;
            this.startCol = startCol;
            this.endCol = endCol;
            this.common = common;
        }

        @Override
        protected void compute() {
            if ((endRow - startRow) * (endCol - startCol) <= THRESHOLD) {
                for (int i = startRow; i < endRow; i++) {
                    for (int j = startCol; j < endCol; j++) {
                        result[i][j] = 0;
                        for (int k = 0; k < common; k++) {
                            result[i][j] += matrixA[i][k] * matrixB[j][k];
                        }
                    }
                }
            } else {
                int midRow = (startRow + endRow) / 2;
                int midCol = (startCol + endCol) / 2;

                invokeAll(
                        new MultiplyTask(matrixA, matrixB, result, startRow, midRow, startCol, midCol, common),
                        new MultiplyTask(matrixA, matrixB, result, startRow, midRow, midCol, endCol, common),
                        new MultiplyTask(matrixA, matrixB, result, midRow, endRow, startCol, midCol, common),
                        new MultiplyTask(matrixA, matrixB, result, midRow, endRow, midCol, endCol, common)
                );
            }
        }
    }
}
